### DISCUSSION TITLE
ValueError: <class 'list'> has no matching SQLAlchemy type

### DISCUSSION THREAD

**Discussion by mockingjet** [fastapi/sqlmodel#995] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlmodel import Field, Relationship, SQLModel

class Role(SQLModel, table=True):
    name: str = Field(index=True, unique=True)
    permissions: list["Permission"] = Relationship(
        back_populates="roles",
        link_model=RolePermissionLink,
    )

class Permission(SQLModel, table=True):
    name: str = Field(index=True, unique=True)
    roles: list[Role] = Relationship(
        back_populates="permissions",
        link_model=RolePermissionLink,
    )
```

### Description

Either `permissions: list["Permission"]` or `roles: list[Role]` is not working and emits this error:

> File "/usr/local/lib/python3.12/site-packages/sqlmodel/main.py", line 615, in get_sqlalchemy_type
    raise ValueError(f"{type_} has no matching SQLAlchemy type")
ValueError: <class 'list'> has no matching SQLAlchemy type

### Operating System

Linux

### SQLModel Version

0.0.18

### Python Version

3.12.4

**Comment by ixycoexzckwpmlcu** [comment, marked as answer]
I've tried to reproduce your example but couldn't. If I run your example verbatim, it errors with `NameError: name 'RolePermissionLink' is not defined`[^1]. If I change your example to the following:
```py
from sqlmodel import Field, Relationship, SQLModel

class RolePermissionLink(SQLModel, table=True):
    role_id: str = Field(foreign_key="role.id", primary_key=True)
    permission_id: str = Field(foreign_key="permission.id", primary_key=True)

class Role(SQLModel, table=True):
    id: int = Field(primary_key=True)
    name: str = Field(index=True, unique=True)
    permissions: list["Permission"] = Relationship(
        back_populates="roles",
        link_model=RolePermissionLink,
    )

class Permission(SQLModel, table=True):
    id: int = Field(primary_key=True)
    name: str = Field(index=True, unique=True)
    roles: list[Role] = Relationship(
        back_populates="permissions",
        link_model=RolePermissionLink,
    )
```
... then it works fine (no console output).

[^1]: Full traceback at https://gist.github.com/ixycoexzckwpmlcu/75e0012542f39f8231f0b3c72e280ca4

**Reply by mockingjet** [reply]
Hi @ixycoexzckwpmlcu, thanks for testing. I have checked multiple times and my code is the same as yours. Can you share which database and which version of sqlmodel you used?

I am using postgres and sqlmodel@0.0.18

**Reply by mockingjet** [reply]
Ah... I read this and now I know I can only declare relationship in **Table Model** https://sqlmodel.tiangolo.com/tutorial/fastapi/teams/#relationship-attributes
I shouldve not put relationship field in BaseModel
